app/api/query/stream/route.ts

```diff
@@ -0,0 +1,100 @@
+import { NextRequest, NextResponse } from 'next/server';
+import { queryChatStream, queryCompletionStream } from 'lib/query';
+// Separators recommended from
+// https://stackoverflow.com/questions/6319551/whats-the-best-separator-delimiter-characters-for-a-plaintext-db-file
+const JSON_SEPARATOR_CODE_POINT = parseInt('1d', 16);
+const CONTENT_SEPARATOR_CODE_POINT = parseInt('1e', 16);
+
+type ChatResponse = {
+  choices: Array<{
+    delta: {
+      content: string;
+    };
+  }>;
+};
+
+type InfoContext = {
+  context: Array<object>;
+};
+
+function iterableToStream(
+  iterable: AsyncIterable<string | ChatResponse>,
+  info: InfoContext,
+  isChat: boolean
+) {
+  const encoder = new TextEncoder();
+  return new ReadableStream({
+    async pull(controller) {
+      for await (const value of iterable) {
+        const chunk = isChat ? (value as ChatResponse).choices[0].delta.content : value;
```

Comment:
[for chat] Doesn't this need to check that `finish_reason` is null and break otherwise?

Something to consider: we currently send the "documents" back after the chat/completion stream finishes, but we have those values before those streams even begin. You could either send them first, or better yet, send them in parallel.

Lastly, `isChat` is not needed. There are two possible values this function accepts (from the `iterable`), one of which is `string`. So you can use that to switch on behavior which also has the benefit of giving you the correct typing, eliminating the need to cast with `as`. For example:

```ts
const chunk: string = typeof value === 'string' ? value : value.choices[0].delta.content!;
controller.enqueue(encoder.encode(chunk));
```

Reply:
>[for chat] Doesn't this need to check that finish_reason is null and break otherwise?

I'm not sure. It works as is, so maybe it's not necessary?

>Something to consider: we currently send the "documents" back after the chat/completion stream finishes, but we have those values before those streams even begin. You could either send them first, or better yet, send them in parallel.

Yep, that was something I wanted to do, but didn't get to. 

> Lastly, isChat is not needed 
Will update with the type switching, that's better and removes signature complexity. Thanks!